49 implement post processing

miklasr opened the pull request:
# Details

- Add implementation for post processing
- Add corresponding tests
- Rework the testing setup
- Refactor internal data structures to remove repetitive code
    - Make certain areas of code less error-prone

# Checklist

## Code
- [x] I have added some tests
- [x] I have run the tests locally to ensure they all pass by running `./test.sh`

## Documentation
- [x] I have updated documentation appropriately (GitHub Pages)
- [x] I have tested the updated GitHub Pages locally to ensure the changes to documentation render correctly (see this [Guidance](https://ese-peasy.github.io/PosturePerfection/guidance.html) page)

Closes #49

miklasr committed:
Create main.yml (#1)

* Created structure for GitHub Pages

* Update config for using remote theme

* Initial project structure (#10)

* Initial setup for using C++ and CMake

* Add initial Doxygen setup and build

For:
Initialise project structure (C++ placeholders) #9

* Create main.yml

* Update main.yml

Add `sudo`

Co-authored-by: Ashwin-MJ <ashwinmj@gmail.com>
Co-authored-by: Ashwin Maliampurakal <45148261+Ashwin-MJ@users.noreply.github.com>

miklasr committed:
Create main.yml (#1)

* Created structure for GitHub Pages

* Update config for using remote theme

* Initial project structure (#10)

* Initial setup for using C++ and CMake

* Add initial Doxygen setup and build

For:
Initialise project structure (C++ placeholders) #9

* Create main.yml

* Update main.yml

Add `sudo`

Co-authored-by: Ashwin-MJ <ashwinmj@gmail.com>
Co-authored-by: Ashwin Maliampurakal <45148261+Ashwin-MJ@users.noreply.github.com>

miklasr committed:
Add cpp linter

miklasr committed:
Add cpp linter

Commit:
Remove old files

These files keep persisting and need to be deleted on master

Commit:
Remove old files

These files keep persisting and need to be deleted on master

Commit:
Initial additions for implementing post processor

Commit:
Add IIR to post processor

Commit:
Add docstrings for IIR header

Commit:
Start work on implementing post processor

Commit:
Merge branch 'main' into 49-implement-post-processing

Commit:
Updated CMakeLists

Commit:
Update CMake for unit testing

Commit:
Remove outdated reference to doubler

Commit:
Add implementation for post processor and tests

Commit:
Remove outdated Boost define

Commit:
Update to make testing post processor possible

Commit:
Update to use vectors and update tests

Commit:
Convert structs to C++ style

Commit:
Add test for post processor

Commit:
Refactor internal structures

- Add a new header for internal structures
- Reduce repetitive code by making results iterable

miklasr commented:
**Note**: The current `mian()` doesn't use my post processor, but the module has been tested with unit testing. I think we need to include multi-frame inputs, before the post-processor can be properly integration tested.

Ashwin-MJ commented on src/iir.cpp:
```diff
@@ -0,0 +1,62 @@
+  };
+}
+
+float IIR2ndOrderFilter::run(float x) {
+  float output = this->nodes.b1 * this->nodes.tap1;
+  x = x - (this->nodes.a1 * this->nodes.tap1);
+  output = output + (this->nodes.b2 * this->nodes.tap2);
+  x = x - (this->nodes.a2 * this->nodes.tap2);
+  output = output + x * this->nodes.b0;
+  this->nodes.tap2 = this->nodes.tap1;
+  this->nodes.tap1 = x;
+  return output;
+}
```
Had to look up my DSP notes for this, looks right to me! Nice stuff 😄

Ashwin-MJ commented on src/iir.h:
```diff
@@ -0,0 +1,129 @@
+  /**
+   * @brief Construct a new `IIRFilter` object
+   *
+   * Passing empty `SmoothingSettings` disbales the IIR filter.
```
*Minor*: ~disbales~ -> disables

miklasr replied:
Fixed

Ashwin-MJ commented on src/intermediate_structures.h:
```diff
@@ -0,0 +1,85 @@
+/**
```
Big fan of this change, makes the code **so** much nicer! 👍 

Only thing is perhaps adding a bit of documentation to the undocumented structs/arrays. I don't mind accepting this PR and then adding this in myself, but I think there should be some explanation why we have two `Coordinate` structs)

Ashwin-MJ replied:
(This was meant to be a comment for the whole file)

miklasr replied:
Thanks, I've updated the documentation and tried to make the distinction between the coordinate types clearer.

Ashwin-MJ commented on src/post_processor.cpp:
```diff
@@ -0,0 +1,73 @@
+  }
+
+  body_part_index = BodyPartMin;
+  for (auto& body_part : results.body_parts) {
+    body_part.x = this->iir_filters.at(body_part_index).run(body_part.x);
+    body_part.y = this->iir_filters.at(body_part_index + 1).run(body_part.y);
+    body_part_index += 2;
+  }
```
Not sure if I'm being silly, but I don't entirely understand how this works?

miklasr replied:
I'll comment this to make it clearer

miklasr replied:
The loop has been merged with the other one in the file. I've also added comments that hopefully make it clearer. Let me know if you think I should make it even clearer or make the use of multiple filters more explicit in some other way.

Ashwin-MJ commented on src/iir.h:
```diff
@@ -0,0 +1,129 @@
+
+#include <vector>
+
+#define COEFFS_PER_IIR2ndOrderFilter 6  //  Coefficients per filter stage
```
Do we use this anywhere?

miklasr replied:
Not used; removed

Ashwin-MJ reviewed (commented):
Awesome work, all looks really great! Moving to `intermediate_structures.h` should make things much better to work with. The tests look great too!

A few very minor comments, mainly for my own understanding.

I will run this on my machine too and let you know how it goes

Commit:
Remove unused define

As per https://github.com/ESE-Peasy/PosturePerfection/pull/55#discussion_r583788409

Commit:
Update main to use post processor

As discussed with @Ashwin-MJ

Ashwin-MJ commented:
Built and passed on my machine using `./test.sh` 🎉

Commit:
Improve documentation

miklasr commented on src/post_processor.cpp:
```diff
@@ -0,0 +1,73 @@
+
+  // Convert to `Inference::Coordinate`
+  int body_part_index = BodyPartMin;
+  for (auto body_part : inference_core_output.body_parts) {
+    results.body_parts[body_part_index] =
+        Coordinate{body_part.x, body_part.y,
+                   (body_part.confidence > this->confidence_threshold)
+                       ? Status::Trustworthy
+                       : Status::Untrustworthy};
+    body_part_index++;
+  }
+
+  body_part_index = BodyPartMin;
+  for (auto& body_part : results.body_parts) {
+    body_part.x = this->iir_filters.at(body_part_index).run(body_part.x);
+    body_part.y = this->iir_filters.at(body_part_index + 1).run(body_part.y);
+    body_part_index += 2;
```
The filtering should be done **before** the `Trustworthy`/`Untrustworthy` label is applied. I'll fix this.

miklasr replied:
I've combined the loops and made sure filtering is done first now. At the moment it doesn't matter, but if we do try out filtering confidence values in time (not sure yet if that would make sense) the change is now trivial.

Commit:
Combine two iterations into single loop

Also ensure IIR filtering done before classification of trustworthiness

Commit:
Make test output verbose

Commit:
Improve comments help distinguish coordinate types

Also minor spelling fix

miklasr commented:
I've made some changes following the comments and I think it should definitely be good to merge now.

C-Begley commented on src/CMakeLists.txt:
```diff
@@ -2,6 +2,14 @@ set(CMAKE_CXX_STANDARD 11)
 
+
+set(LIBSRC
+  doubler.cpp
```
Should we give this a more meaningful name or even just remove this?

miklasr replied:
Doubler has been removed. I forgot to do that so thanks for spotting!

Ashwin-MJ reviewed (approved):
Changes look good to me!

Commit:
Remove example doubler files
